### ISSUE TITLE
Add support for policy "audit" mode vs "enforcing" mode

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#9580]
Currently if you write policies in Cilium, you must get your policies correct the first time, otherwise you will start dropping traffic.

Various projects like SELinux have a notion of an "audit" mode where the policy is put in place, but traffic is not denied; rather, if any traffic *would* be denied, then audit logs are generated to inform the user that their current policy is not sufficient to cover the set of operations that are occurring.

We could introduce such a notion into the BPF datapath, and configure it such that if this mode is enabled and the traffic was expected to be dropped, then we generate a new monitor message type that indicates that the traffic would be dropped by policy but is allowed due to the audit mode.

The mode for this flag should be set to "enforcing" by default unless otherwise configured by the user.

We may consider allowing the user to configure this via k8s annotations similar to the visibility mode (#9088). We would likely also need a global flag to allow or disable the "audit" mode, so that the cluster administrator has global control over whether this feature is allowed to be used. It should default to "enforcing" mode.

Alternatively we could consider tying it closer to the CNP spec or just have a single global flag so that the administrators responsible for policy or deployment are the ones who configure this mode.

Open question is what granularity is useful here - per node? per namespace? per pod?

**Comment by danwent** [comment]
A big +10 for this issue, as I've heard it mentioned by many users. 

One idea would be to have an annotation on the CNP itself, with the implication that all endpoints that are selected by a CNP with that annotation are in warn mode.  A common case would be to have a default deny policy that selects all endpoints in the namespace, and if that policy had the annotation, it would just be audit mode.  But it would also be possible to create more fine-grained policy that just select a subset or single endpoint, and use audit mode only for those endpoints.   And in the future, when we have global policy, it would even be possible to use annotation on a global policy as well.

We could then have "cilium endpoint list" indicate the status of the endpoint.   I guess this raises the question of whether this annotation can indicate whether audit mode is applied for ingress/egress/both (given that existing enforcement status for ingress/egress is separate on a per endpoint basis)

**Comment by joestringer** [comment]
Global daemon mode for policy audit is now available on latest (to be part of v1.8) via #9970.

**Comment by nyrahul** [comment]
Hello, Would like to understand if there has been any discussion of this front post the global daemon mode policy audit availability.

It may be problematic for the users to apply a daemon wide audit mode when only a single policy is to be staged. If a single policy is to be staged in production and if this requires enabling daemon wide audit mode, then it may be a security risk since it essentially enables traffic across all flows on that node.

[Calico supports staging](https://www.tigera.io/features/staged-policies/) a single policy i.e, you can create and stage a policy and observe traffic impact because of that policy. This is an important one since it allows the users to gauge the impact of new policy addition without impacting other committed/enforced policies.

Would like to understand if this was under consideration and if any technical discussions had happened in the context elsewhere. Thanks.

**Comment by joestringer** [comment]
Hi @nyrahul , I'm not aware of any ongoing discussions. If you're interested in tackling this problem, we'd be happy to discuss a proposal and this issue would be as good a place as any to discuss (though we can also discuss via Slack or Google Docs if that's more convenient).

**Comment by nyrahul** [comment]
> Hi @nyrahul , I'm not aware of any ongoing discussions. If you're interested in tackling this problem, we'd be happy to discuss a proposal and this issue would be as good a place as any to discuss (though we can also discuss via Slack or Google Docs if that's more convenient).

Sure, maybe I ll first prepare some initial analysis and then seek inputs on the proposal (using slack/google docs, like you mentioned). Thanks for the response.

**Comment by mazzy89** [comment]
@joestringer we put black on white the first proposal about the audit mode support at the level of the Cilium Network Policy and we are keen to grab this and working actively on this. I have shared the Google doc proposal with @joestringer.

**Comment by pchaigno** [comment]
> @joestringer we put black on white the first proposal about the audit mode support at the level of the Cilium Network Policy and we are keen to grab this and working actively on this. I have shared the Google doc proposal with @joestringer.

@mazzy89 That's great! Could you share the link here? Might also be worth sharing in Slack #development.

**Comment by mazzy89** [comment]
Sure @pchaigno. Here the [document](https://docs.google.com/document/d/13mC0PeZRKLx2lwu6cJmQ3VftSJ8sqJLokZz8XuYpUbk/edit?usp=sharing)

I will probably shout it today during the weekly dev meeting just for FYI. We'll discuss this in the upcoming days.

**Comment by ArthurChiao** [comment]
Hi @mazzy89, this feature sounds great! May I know the progress of it?

**Comment by mazzy89** [comment]
Hi @ArthurChiao above you will find the proposal. It's marked at the moment as a draft. Cilium's folks have to review it.

**Comment by mazzy89** [comment]
After some time and discussions, we are now ready to move with the implementation of this feature. The proposal has been marked now as final and it is available to the aforementioned link.

We are going to start with the implementation of the audit mode per Cilium Network Policy. We'll consider separately in a different issue the display of the verdicts.

At the time of this writing, the log verdicts are available at the level of the Cilium socket. Separately we want to enrich Hubble flow with information about the Cilium Network Policy where the audit mode is applied.

**Comment by chancez** [comment]
@mazzy89 any progress on this? It's been a while and there's been a few questions in the Cilium slack that would indicate this could be useful for others.

**Comment by jimethn** [comment]
I think this should be closed, I think the feature exists: https://docs.cilium.io/en/stable/security/policy-creation/

**Comment by puja108** [comment]
Stumbled upon this as we had a related use case where deny-all is enabled on some namespaces and we would like to move to have global deny-all, for that we would need to have audit mode for everything but the namespaces that are already "compliant" to get the rest of workloads compliant. A namespace by namespace approach, i.e. being able to enable audit mode per namespace or per CNP would help a lot.

From what I see the current feature is either global audit mode or per endpoint, so the use case is a bit tedious to set up, basically having to cycle through all endpoints and set them to audit mode. Other use cases like introducing a single new policy would also not be supported.

**Comment by chancez** [comment]
> I think this should be closed, I think the feature exists: https://docs.cilium.io/en/stable/security/policy-creation/

That's an agent level config option, effecting all policies for that node, this issue is about a policy level option.

**Comment by kgtw** [comment]
Is there any progress being made on this issue? I tried to look at the google document linked by @mazzy89 but it appears to have been deleted.

> Sure [@pchaigno](https://github.com/pchaigno). Here the [document](https://docs.google.com/document/d/13mC0PeZRKLx2lwu6cJmQ3VftSJ8sqJLokZz8XuYpUbk/edit?usp=sharing)
> 
> I will probably shout it today during the weekly dev meeting just for FYI. We'll discuss this in the upcoming days.

**Comment by bowei** [comment]
There is some potentially relevant discussion here:

https://github.com/kubernetes-sigs/network-policy-api/pull/276

from upstream K8s.

**Comment by nathanjsweet** [comment]
Few will agree, but I want to express my disapproval of this feature. This feature is harmful to network security and stability.

This feature is most beneficial to large cluster operators. The feature promises assurance that they can "stage" a network policy, look through their logs, and assume there is no problem if the logs say so. This is not a professional way to stage network policies. Especially in larger clusters, there is no way of knowing if enough time has elapsed in a staged audit run of a network policy that all potential network connections in the cluster are unaffected. This approach to network security is blind to the actual state of the cluster. This feature is a rube-goldberg device that does not actually give any reasonable assurance that a policy won't break a cluster, especially in the long term. More deterministic approaches to avoiding policy-networking collision should be used.

**Comment by bowei** [comment]
I agree that this is not ideal, but as I understand it, it is a common workflow for admins who want to secure an application that wasn't secured previously to create a number of audit policies to see what the impact is and whether or not the admin's understanding of the application's dependencies is correct before turning the policies on in enforcement mode.

Our inner computer scientists may blanch at such a procedure as fundamentally flawed, but it is common enough that we can't discount the user need...

**Comment by RajatGarga** [comment]
@nathanjsweet 
> There are other ways to solve the issue of checking network policy collisions.
Could you elaborate on this?

Currently, we face challenges in assessing the immediate impact of a network policy. Our primary options include applying all policies in audit mode or testing them in a separate staging cluster. Both approaches have limitations. For example, enabling audit mode globally can pose security risks, as it permits traffic across all flows on the node. Meanwhile, staging policies in a separate cluster is not always practical and may not fully capture potential consequences in production.

While I acknowledge that staging policies won't guarantee there are no unintended consequences, it's important to have a mechanism to introduce policies gradually. The reliability of a staged policy depends on the operator's expertise and familiarity with the cluster. Dismissing this option outright seems restrictive, particularly when the audit mode for the daemon already exists.

Looking forward to hearing your thoughts on alternative solutions!
